src/deepsparse/timing/timing_builder.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from typing import Dict
+
+from pydantic import BaseModel
+
+from deepsparse.timing.timing_schema import InferenceTimingSchema
+
+
+__all__ = ["TimingBuilder"]
+
+
+class TimingBuilder:
+    """
+    This object aggregates the durations
+    (time deltas in seconds) of various components of the inference
+    pipeline.
+    Once all the desired time deltas are aggregated, the
+    object may build the data structure that summarizes the
+    collected information.
+
+    Example flow:
+
+    ```
+    builder = TimingBuilder()
+    builder.initialize()
+
+    builder.start("total_inference")
+
+    builder.start("pre_process")
+    do_something()
+    builder.stop("pre_process")
+
+    builder.start("engine_forward")
+    do_something()
+    builder.stop("engine_forward")
+
+    builder.start("post_process")
+    do_something()
+    builder.stop("post_process")
+
+    builder.stop("total_inference")
+    ...
+    summary = builder.build()
+    ```
+    The object may time the duration of an arbitrary number
+    of events (phases), but the names of phases need to
+    be consistent with the `schema` argument passed to
+    the build() method
+    """
+
+    def __init__(self):
+        self.initialized = False
+        self._start_stop_times = {}
```

Comment:
Can we get this split into `self.start_times = {}` and `self.stop_times = {}`? I think it'd make it slightly easier to validate that something has started, and slightly easier to read